solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/ModelQuery.java

```diff
@@ -146,14 +159,84 @@ public SolrQueryRequest getRequest() {
+    boolean newCode = true;
+    if (!newCode){
+      this.featuresRequested = true;
+      long  start = System.currentTimeMillis();
+      final FeatureWeight[] allFeatureWeights = createWeights(meta.getAllFeatures(),
+          searcher, needsScores);
+      final FeatureWeight[] modelFeaturesWeights = createWeights(meta.getFeatures(),
+          searcher, needsScores);
+      long  end = System.currentTimeMillis();
+      log.info("In ModelQuery.java::createWeight() OLD createWeight time:{} ms", (end-start));
+      return new ModelWeight(searcher, modelFeaturesWeights, allFeatureWeights);
+    }
+    else{      
+      long  start = System.currentTimeMillis();
+      final Collection<Feature> modelFeatures = meta.getFeatures();
+      final Collection<Feature> allFeatures = meta.getAllFeatures();
+          
+      int modelFeatSize = modelFeatures.size();
+      int allFeatSize = this.featuresRequested ? allFeatures.size() : modelFeatSize;
+
+      final FeatureWeight[] extractedFeatureWeights = new FeatureWeight[allFeatSize];
+      final FeatureWeight[] modelFeaturesWeights = new FeatureWeight[modelFeatSize];
+      createWeightsSelectively(allFeatures,modelFeatures,searcher, needsScores, extractedFeatureWeights,modelFeaturesWeights); 
+      
+      int modelFeatIndex = 0;
+      Normalizer[] modelFeatureNorms = new Normalizer[modelFeatures.size()]; // store the featureNorms for modelFeatures to use later for normalization
+      for (final Feature modelFeature: modelFeatures){
+        modelFeatureNorms[modelFeatIndex++]= modelFeature.getNorm();
+      }
+      
+      long  end = System.currentTimeMillis();
+      log.info("In ModelQuery.java::createWeight() NEW!  createWeight time:{} ms", (end-start));
+      
+      return new ModelWeight(searcher, modelFeaturesWeights, extractedFeatureWeights, modelFeatureNorms);
+    }
   }
-
+  
+  private void createWeightsSelectively(Collection<Feature> allFeatures, 
+      Collection<Feature> modelFeatures,
+      IndexSearcher searcher, boolean needsScores, FeatureWeight[] allFeatureWeights,
+      FeatureWeight[] modelFeaturesWeights) throws IOException {
+    long  start = System.currentTimeMillis();
+    int i = 0, j = 0;
+    final SolrQueryRequest req = getRequest();
+    // since the feature store is a linkedhashmap order is preserved
+    if (this.featuresRequested) {
+      final HashSet<Feature> modelFeatSet = new HashSet<Feature>(modelFeatures);
+      for (final Feature f : allFeatures) {
+        try{
+        FeatureWeight fw = f.createWeight(searcher, needsScores, req, originalQuery, efi);
+        allFeatureWeights[i++] = fw;
+        if (modelFeatSet.contains(f)){  
+          // Note: by assigning fw created using allFeatures, we lose the normalizer associated with the modelFeature, for this reason, 
+          // we store it in the modelFeatureNorms, ahead of time, to use in normalize() and explain()
```

Comment:
We have a ticket to factor out the normalizers from the feature class itself, which could impact this comment (and the normalizer loop above)